test/controllers/helpers/HomeCardGeneratorSpec.scala

```diff
@@ -116,144 +119,136 @@ class HomeCardGeneratorSpec extends BaseSpec {
       implicit lazy val pertaxContext =
         PertaxContext(FakeRequest(), mockLocalPartialRetreiver, injected[ConfigDecorator])
 
-      def taxCalcState: TaxCalculationState
+      implicit lazy val configDecorator = injected[ConfigDecorator]
 
-      lazy val cardBody = c.getTaxCalculationCard(Some(taxCalcState), 2015, 2016)
+      def taxYearRec: TaxYearReconciliation
+
+      lazy val taxCalculationViewModel = TaxCalculationViewModel.fromTaxYearReconciliation(taxYearRec)
+
+      lazy val cardBody = serviceUnderTest.getTaxCalculationCard(Some(taxYearRec))
     }
 
-    "return nothing when called with TaxCalculationUnderpaidPaymentsDownState" in new LocalSetup {
-      val taxCalcState = TaxCalculationUnderpaidPaymentsDownState(2015, 2016)
+    "return nothing when called with Underpaid PaymentsDown reconciliation status" in new LocalSetup {
+      val taxYearRec = TaxYearReconciliation(2015, Underpaid(None, None, PaymentsDown))
 
       cardBody shouldBe None
     }
 
-    "return nothing when called with TaxCalculationUnkownState" in new LocalSetup {
-      val taxCalcState = TaxCalculationUnkownState
+    "return correct markup when called with Balanced status" in new LocalSetup {
+      val taxYearRec = TaxYearReconciliation(2015, Balanced)
 
-      cardBody shouldBe None
+      cardBody shouldBe taxCalculationViewModel.map(taxCalculation(_))
     }
```

Comment:
It seems like there are no tests on TaxCalculationViewModel and they are side tested here, would it make sense to push the view model logic into the HomeGenerator (or something else)

Reply:
I have looked into this suggestion and believe that it may not be a good idea to move the logic from the View model to the HomeCardGenerator as the generator only serves to return HTML under certain circumstances for each type of tile in Pertax. 

Putting that logic there may make the Generator more convoluted, therefore the decision was made to strip the logic out of the view itself and create a model that create objects with the most appropriate structure, which could then be passed to the view to render in a clear and simple way.

Reply:
Sorry, I've probably not explained it properly in my first go. 

The tests on TaxCalculationViewModel appear to live in the TaxCalculationViewSpec, I think it would better to place them in a TaxCalculationViewModelSpec and have the ViewSpec just testing that it can print out any link/url. At the moment the ViewModel tests are tied to the ViewSpec tests so the tests are both testing the correct url/labels are being rendered but what you really care about is that the ViewModel creates the right urls/labels and the ViewSpec is capable of rendering any url/label. ViewSpecs should be really light weight as when you start testing logic in a ViewSpec they can become large and unwieldy due to testing logic + rendering simultaneously. We should always be trying to split UI and logic out into separate independent modules where they share no concerns.